File: src/pat/Advance1101.java
package pat;

import java.util.*;

/**
 * Created by zq on 2016/8/17.
 */
public class Advance1101  {

    private static int[] myList;
    private static boolean[] isPivot;
    private static List<Integer> answer;

    private static final int MAX_NUMBER = 100001;

    public static void main(String[] arg){

        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        myList = new int[n];
        isPivot = new boolean[n];
        answer = new ArrayList<Integer>();

        for(int i=0;i<n;i++) {
            myList[i] = scanner.nextInt();
            isPivot[i] = true;
        }

        int leftMaxValue = myList[0];
        for(int i=0;i<n;i++){
            int a = myList[i];
            if(a>=leftMaxValue) leftMaxValue = a;
            else isPivot[i] = false;
        }

        int rightMinValue = myList[n-1];
        for(int i=n-2;i>=0;i--){
            int a = myList[i];
            if(a<=rightMinValue) rightMinValue = a;
            else isPivot[i] = false;
        }

        for(int i=0;i<n;i++)
            if(isPivot[i])
                answer.add(myList[i]);

        Collections.sort(answer);

        System.out.println(answer.size());

        if(!answer.isEmpty()) System.out.print(answer.get(0));
        for(int i=1;i<answer.size();i++)
            System.out.print(" "+answer.get(i));
        System.out.println();

    }
}

File: src/pat/Advance1101.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int myList[100001];
int isPivot[100001];
vector<int>answer;

int main()
{
    int n;
    cin>>n;
    for(int i=0;i<n;i++){
        cin>>myList[i];
        isPivot[i] = true;
    }

    int leftMaxValue = myList[0];
    for(int i=0;i<n;i++)
    {
        if(myList[i]>=leftMaxValue) leftMaxValue = myList[i];
        else isPivot[i] = false;
    }

    int rightMinValue = myList[n-1];
    for(int i=n-1;i>=0;i--)
    {
        if(myList[i]<=rightMinValue) rightMinValue = myList[i];
        else isPivot[i] = false;
    }

    for(int i=0;i<n;i++)
        if(isPivot[i]) answer.push_back(myList[i]);

    sort(answer.begin(),answer.end());
    cout<<answer.size()<<endl;
    if(!answer.empty()) cout<<answer[0];
    for(int i=1;i<answer.size();i++)
        cout<<" "<<answer[i];

    cout<<endl;
}
